colcon_core/package_descriptor.py

```diff
@@ -114,7 +114,8 @@ def get_recursive_dependencies(
         :raises AssertionError: if a package lists itself as a dependency
         """
         if not isinstance(recursive_categories, Mapping):
-            recursive_categories = defaultdict(lambda: recursive_categories)
+            non_map_categories = recursive_categories
+            recursive_categories = defaultdict(lambda: non_map_categories)
```

Comment:
Reading up on lambda captures, you might be able to do this (pardon the line length):

```suggestion
            recursive_categories = defaultdict(lambda recursive_categories=recursive_categories: recursive_categories)
```

but it's not really better than the current solution.